gcp/appengine/Dockerfile

```diff
@@ -0,0 +1,56 @@
+# Build the Javascript frontend
+FROM node:latest AS FRONTEND3_BUILD
+WORKDIR /build/frontend3
+
+# Install dependencies first for better caching
+COPY gcp/appengine/frontend3/package.json gcp/appengine/frontend3/package-lock.json ./
+RUN npm ci
+
+COPY gcp/appengine/frontend3/webpack.prod.js ./
+COPY gcp/appengine/frontend3/img img
+COPY gcp/appengine/frontend3/src src
+
+RUN npm run build:prod
+
+# Build hugo blogs
+# Use the ci image, since it already built the version of hugo we want from source
+FROM gcr.io/oss-vdb/ci AS HUGO_BUILD
+
+WORKDIR /build/blog
+COPY gcp/appengine/blog ./
+
+RUN hugo -d ../dist/static/blog
+
+# OSV.dev site image
+# Adapted from https://cloud.google.com/run/docs/quickstarts/build-and-deploy/deploy-python-service#writing
+FROM python:3.11-slim
+
+# Allow statements and log messages to immediately appear in the logs
+ENV PYTHONUNBUFFERED True
+WORKDIR /osv/gcp/appengine
+
+# Install Python dependencies
+COPY setup.py Pipfile* README.md /osv/
+COPY osv /osv/osv
+COPY gcp/appengine/Pipfile* ./
+RUN pip3 install -U pipenv==2023.6.12 && python3 -m pipenv install --deploy --system
+
+# Website Python code
+COPY gcp/appengine/*.py ./
+
+# JS/hugo builds
+COPY gcp/appengine/dist/public_keys dist/public_keys
+COPY gcp/appengine/docs docs
+# gcp/appengine/docs/docs/osv_service_v1.swagger.json is a symlink
+COPY docs/osv_service_v1.swagger.json docs/
+
+COPY --from=FRONTEND3_BUILD /build/dist/ dist/
+COPY --from=HUGO_BUILD /build/dist dist/
+
+
+# Run the web service on container startup. Here we use the gunicorn
+# webserver, with one worker process and 8 threads.
+# For environments with multiple CPU cores, increase the number of workers
+# to be equal to the cores available.
+# Timeout is set to 0 to disable the timeouts of the workers to allow Cloud Run to handle instance scaling.
+CMD exec gunicorn --bind :$PORT --workers 1 --threads 8 --timeout 0 main:app
```

Comment:
If I recall correctly, there hadn't been a whole lot of rigour around the suitability of the selection of gunicorn for the serving stack, right?

If https://www.fullstackpython.com/wsgi-servers.html is to be considered authoritative, uWSGI may be the most performant choice. I definitely think we should be running the final choice through https://best.openssf.org/Concise-Guide-for-Evaluating-Open-Source-Software

Reply:
It seems uWSGI is in maintenance mode FYI (at least according to their docs)

Reply:
I think gunicorn is the default appengine choice, so at least this maintains the same behavior as before. Agreed with going through some basic evaluation before moving this to production though.

Reply:
> It seems uWSGI is in maintenance mode FYI (at least according to their docs)

Is that a bad thing? Less changes == less bugs == better reliability (in theory)

Reply:
> Is that a bad thing? Less changes == less bugs == better reliability (in theory)

That's certainly one way to look at it and can be true in some situations. There is some subjectivity about what maintenance mode means depending on the project & maintainers.

Rightly/wrongly, I would usually interpret it as less effort being put into the software in general and possibly delayed response for things that need to be worked on (bugs/security vulns etc)

I think it also means that shiny new protocols and features are less likely to be supported, which can be both good and bad depending on the situation.  Worth evaluating against the criteria you shared to see where it ends up.